ducc/src/object.rs

```diff
@@ -211,16 +224,80 @@ impl<'ducc> Object<'ducc> {
     }
 }
 
-#[derive(Default)]
-pub struct PropertyDescriptor<'ducc, V>
-where
-    V: ToValue<'ducc> {
-    pub enumerable: bool,
-    pub configurable: bool,
-    pub writable: bool,
-    pub value: Option<V>,
-    pub get: Option<Function<'ducc>>,
-    pub set: Option<Function<'ducc>>,
+enum PropertySource<'ducc> {
+    Undefined,
+    Value(Value<'ducc>),
+    GetSet(Function<'ducc>, Function<'ducc>),
+    Get(Function<'ducc>),
+    Set(Function<'ducc>),
+}
+
+pub struct PropertyDescriptor<'ducc> {
+    enumerable: Option<bool>,
+    configurable: Option<bool>,
+    writable: Option<bool>,
+    source: PropertySource<'ducc>
+}
+impl <'ducc> PropertyDescriptor<'ducc> {
+    pub fn new() -> PropertyDescriptor<'ducc> {
+        PropertyDescriptor {
+            enumerable: None,
+            configurable: None,
+            writable: None,
+            source: PropertySource::Undefined
+        }
+    }
+
+    /// Whether this property shows up during enumeration of the
+    /// properties on the corresponding object.
+    /// 
+    /// Defaults to `false`
+    pub fn enumerable<'a>(&'a mut self, b: bool) -> &'a mut PropertyDescriptor<'ducc> {
```

Comment:
Shouldn't these functions also be `self`-consuming like `value`/`getter`/etc.?

Reply:
Hmm maybe. I followed the Rust builder pattern guide on non-consuming patterns: https://doc.rust-lang.org/1.0.0/style/ownership/builders.html#non-consuming-builders-(preferred) but I'm having some problems actually trying to use the one-line builders now. I'll try to change it to something more usable.

Reply:
Ah OK. I'm indifferent between consuming vs. non-consuming but if you're gonna do non-consuming then `define_prop` should take a `&PropertyDescriptor` instead of a `PropertyDescriptor`

Reply:
I ended up going all in on consuming. I don't think complex configuration will be that often needed for property descriptors anyway.